### ISSUE TITLE
SDCard mounting

### ISSUE THREAD

**Issue by mchobby** [RC2014Z80/picoterm#35]
Adding SDCard support on PicoTerm.

Base work on no-OS-FatFS-SD-SPI .
See the following ressource:
* [Raspberry Pi Pico (RP2040) SD Card Example with MicroPython and C/C++](https://www.digikey.be/en/maker/projects/raspberry-pi-pico-rp2040-sd-card-example-with-micropython-and-cc/e472c7f578734bfd96d437e68e670050)
* [no-OS-FatFS-SD-SPI-RPi-Pico](https://github.com/carlk3/no-OS-FatFS-SD-SPI-RPi-Pico) : Simple library for SD Cards on the Pico

Currently under progress ...

**Comment by abaffa** [comment]
Is it already using the GPIO expander or the sdcard module is connected directly to the rp2040? 
Could u send a pic of showing the connections?

**Comment by mchobby** [comment]
I do use the recommandation of Steven documented here:

[PicoTerm connector](https://github.com/RC2014Z80/picoterm/blob/main/docs/picoterm-conn.md)

As soon I do read the FAT, I will update documentation with wiring and "compiling" documentation with the task to do.

Dominique

**Comment by RFC2795** [comment]
I received the PCBs for the SD Card Backpack over the Christmas holidays, as you can see from the photos attached.

I feel like this is probably already out of date without the buzzer or IO expansion, so I think it might be best for me to redesign the whole module to include the various updates discussed here.

(In the meantime, would you like me to send you both one of these for your own use?)

What do you guys think?

Thanks

Spencer

![IMG_20221230_150346](https://user-images.githubusercontent.com/20264032/214142904-cb7fd228-2e53-49f4-9bce-a2cd11a2ddb7.jpg)
![IMG_20221230_150356](https://user-images.githubusercontent.com/20264032/214142920-2c49584f-5f34-45f9-a9f5-cfc97ffc88a2.jpg)
![IMG_20221230_150451](https://user-images.githubusercontent.com/20264032/214142925-599ec749-9bd3-49f1-9ee2-0315cf980be5.jpg)
![IMG_20221230_150456](https://user-images.githubusercontent.com/20264032/214142927-0cab9f1a-91a6-4221-b8c2-007d2dee9bf8.jpg)
![IMG_20221230_150753](https://user-images.githubusercontent.com/20264032/214142928-3f6a12b0-a7ba-494c-9035-804a3dc01e46.jpg)

**Comment by mchobby** [comment]
This is very kind... Thanks.
I like the placement behind the board.

In your new design, could foreseen a footprint for the PCA GPIO expander (as optionnal) ? 

Dominique

**Comment by mchobby** [comment]
BAD NEWs for the first attempt.

I was really expecting good progressing with the no-OS-FatFS-SD-SPI-RPi-Pico library.
Unfortunately that library requires a SPI bus to work properly... unfortunately, we don't have a complete bus available :-(
... just starting the SPI0 do hang software on the Pico.
For tonight, I will revert the work.

I will tray another approach based on [pico_sd_card](https://github.com/raspberrypi/pico-extras/tree/master/src/rp2_common/pico_sd_card)  and its [sd_test](https://github.com/raspberrypi/pico-extras/tree/master/test/sd_test) code
Apparently this can work in 1bit mode but required 3 consecutives GPIO (since it uses PIO). So I'm not very optimistic but "one thing at the time".

**Comment by abaffa** [comment]
@RFC2795 Thanks, This is very kind and would help with development :) 

@mchobby is right. It would be nice to add a GPIO for future expansions.

**Comment by mchobby** [comment]
> @RFC2795 Thanks, This is very kind and would help with development :)
> 
> @mchobby is right. It would be nice to add a GPIO for future expansions.

There are still two GPIOs remaining available on the [PCA9536 schematic](https://github.com/RC2014Z80/picoterm/blob/main/docs/picoterm-conn.md) if you are using it. And you can still add another GPIO expansion as needed (PCA9354 bit 8bit or MCP23017 with 16 bits). So we are relax with this.
The only concerns about that is the relatively low speed of I2C (400 KBits/sec).

Note: when reading the Pico-Implementation guide, I noticed that demo schematics requires 5 contiguous PIN block (for PIO). GP18 to GP22 which are used for other activities.

![Sélection_150](https://user-images.githubusercontent.com/2701466/215142418-81dd1888-f354-4f53-8143-4e1a9ab66f22.png)

Next week-end, I will made additional check/elaboration to:

- either free-up an SPI bus somewhere or 
- either the 5 Pin block to attach the SDCard or 
- either using the one-bit mode SDCard routine (over 3 contiguous pin).

**Comment by mchobby** [comment]
@RFC2795 
@abaffa 
Here is the result of additional check/elaboration around SDCard in the current situation:
![picoterm-sd-options](https://user-images.githubusercontent.com/2701466/215751017-befe54c0-84f2-4243-8608-c7ba687a607a.png)
### Option 1: 4 bits data bus = Reject
It is the solution proposed in the "RP2040 Hardware Design" as followed by @RFC2795 (Spencer).
As seen on the Wiring, I2C bus and UART1 must be deinit before starting PIO engine to access the SD and vice-versa.
I'm not comfortable with such approach. UART1 deinit also means stopping communication with the retro-computer (no buffering on the RX in such case).

### Option 2: 1 bits data bus = Reject 
The "RP2040 Hardware Design" also mentioned the ability of using 1 bit data bus. This requires 3 consecutive GPIO for PIO the SDCard with a state machine as stated in the documentation (or the demonstrating code). 
Unfortunately we don't have such without deinit the UART1. So this is also a failure.

### Option 3: Accessing SDCard in SPI mode = Reject
To get an hardware SPI, we have to wire the DAT0 to GPIO 16 which is also VGAVS (VGA Vertical Sync). This would also requires to validate or create a library for accessing the SDCard.
So when talking to SD, we must stop the VGA rendering thread, stop StateMachine of VGA, init SPI, dial the SD, deinit SPI, reinit GP16, restart VGA statemachine & VGA rendering thread.
For sure, this will be quite glitchy and maybe difficult to synchronize.

### Option 4: I2C-to-SPI bridge (to explore)
Some research lead me to the [NXP SC18IS602B](https://www.nxp.com/products/peripherals-and-logic/signal-chain/bridges/ic-bus-to-spi-bridge:SC18IS602B) which allows to bring a SPI bus over an existing I2C bus.
![Sélection_174](https://user-images.githubusercontent.com/2701466/215757864-e78f34e4-ac5f-424a-ab5a-b514a4541348.png)
I also find this instructive document [http://www.dejazzer.com/ee379/lecture_notes/lec12_sd_card.pdf](http://www.dejazzer.com/ee379/lecture_notes/lec12_sd_card.pdf) about accessing a SD in SPI mode.
This is still a very low level access and library must be coded for FAT access (maybe reusing the FatSD of Arduino Lib?).

[EDIT] ordered an evaluation board for quick testing

### Option  5: TinyUSB - flash drive reader = REJECT
_Fev 2023: Using USB based storage is not really compatible with config file storage._

TinyUSB can support USB Hub (read it somewhere but not yet supported in the PicoTerm code).
Among the peripheral supported by TinyUSB, can it read an USB Stick (USB Flash drive) when plugged ?
In such case... no need for SDCard support.

[EDIT] after some research, I did locate the [tinyusb/examples/host/msc_file_explorer/src/msc_app.c](https://github.com/hathach/tinyusb/blob/master/examples/host/msc_file_explorer/src/msc_app.c) example where TinyUSB do mount a USB Storage and make IO access to him.
The example also includes following functions:
```
void cli_cmd_cat(EmbeddedCli *cli, char *args, void *context);
void cli_cmd_cd(EmbeddedCli *cli, char *args, void *context);
void cli_cmd_cp(EmbeddedCli *cli, char *args, void *context);
void cli_cmd_ls(EmbeddedCli *cli, char *args, void *context);
void cli_cmd_pwd(EmbeddedCli *cli, char *args, void *context);
void cli_cmd_mkdir(EmbeddedCli *cli, char *args, void *context);
void cli_cmd_mv(EmbeddedCli *cli, char *args, void *context);
void cli_cmd_rm(EmbeddedCli *cli, char *args, void *context);
```

### Option  6: reducing VGA bit, relocating UART (to explore)
As suggested by RFC2795 (_freb 2023_), LSB bit (and buttons) can be removed from VGA generator. Still providing with 4096 colors (12 bits). Then UART can be relocated to free-up GP20 & GP21 for SD card (SPI mode).

**Comment by RFC2795** [comment]
The way I envisage a SD card being used is to hold things like fonts and sprites, and maybe user definable macros that can be called with function keys, or config data.  So although a USB drive could be used the same, it will be messier.

If both Option 1 and Option 2 need the UART pins, then maybe the colour depth could be reduced to free up some pins to allow the UART to be moved elsewhere.  The hardware design from Raspberry Pi has 15 bits (and 15 pins) of data, which gives 32768 colours available on the screen at one time.  Whilst this is great, and can be used for demo scene stuff, most CP/M software that uses colours will only use a few primary colours.  The LSB for each colour is currently shared with the A/B/C buttons, but if the LSB each colour was dropped, it would still be a 12 bit colour depth, which gives 4096 colours and frees up 3 pins to move the UART around.

How does that sound?  

I am currently looking in to a new hardware design for this, to incorporate all the modifications which have been discussed here, and it would be nice to put everything together on a single PCB.

**Comment by mchobby** [comment]
@RFC2795 
1) The LSB and buttons could, indeed be removed. I will check feasibility of Pin assignment based on your suggestion and give additional feedback. 

2) I ordered the [NXP SC18IS602B](https://www.nxp.com/products/peripherals-and-logic/signal-chain/bridges/ic-bus-to-spi-bridge:SC18IS602B) I2C-to-SPI bridge (for option 4). Currently waiting for delivery here in Belgium.

**Comment by dquadros** [comment]
Not sure if you want to make this big change, but if you are thinking of reducing the number of bits per pixel there is the picovga project (https://breatharian.eu/hw/picovga/index_en.html). It uses only 8 bits per pixel but supports 80 columns color text.
This should release enough pins for connecting the SD card to one of the two SPI ports.

Enviado do Yahoo Mail no Android 
 
  Em seg., 6 6e fev. 6e 2023 às 18:32, ***@***.***> escreveu:   


The way I envisage a SD card being used is to hold things like fonts and sprites, and maybe user definable macros that can be called with function keys, or config data. So although a USB drive could be used the same, it will be messier.

If both Option 1 and Option 2 need the UART pins, then maybe the colour depth could be reduced to free up some pins to allow the UART to be moved elsewhere. The hardware design from Raspberry Pi has 15 bits (and 15 pins) of data, which gives 32768 colours available on the screen at one time. Whilst this is great, and can be used for demo scene stuff, most CP/M software that uses colours will only use a few primary colours. The LSB for each colour is currently shared with the A/B/C buttons, but if the LSB each colour was dropped, it would still be a 12 bit colour depth, which gives 4096 colours and frees up 3 pins to move the UART around.

How does that sound?

I am currently looking in to a new hardware design for this, to incorporate all the modifications which have been discussed here, and it would be nice to put everything together on a single PCB.

**Comment by dylanhall** [comment]
If you're considering more radical changes to the board then one option is to use one of the other RP2040 boards that expose more pins. For example the PGA2040 (https://shop.pimoroni.com/products/pga2040) or the RP2040 Stamp (https://www.tindie.com/products/arturo182/rp2040-stamp/). I think they make an extra 3 or 4 pins available.

**Comment by mchobby** [comment]
> @RFC2795
> 
>     1. The LSB and buttons could, indeed be removed. I will check feasibility of Pin assignment based on your suggestion and give additional feedback.
> 
>     2. I ordered the [NXP SC18IS602B](https://www.nxp.com/products/peripherals-and-logic/signal-chain/bridges/ic-bus-to-spi-bridge:SC18IS602B) I2C-to-SPI bridge (for option 4). Currently waiting for delivery here in Belgium.

@RFC2795 
I identified the PIXEL_RSHIFT, COLOR_PIN_BASE, COLOR_PIN_COUNT in the PICO_EXTRA/scanVideo .
I will order a new PicoTerm board to test 12 bit rendering from pin 3. So uart_0 will be released on pin 0 & 1  and we would release GP20 & GP21 for SDCard access.

**Comment by RFC2795** [comment]
@mchobby Send me an email (spencer at z80kits dot com) with your address and I'll post a PicoTerm board out to you

**Comment by mchobby** [comment]
> @mchobby Send me an email (spencer at z80kits dot com) with your address and I'll post a PicoTerm board out to you

Hi Spencer,
I just send an email because of shipping issue to Belgium :-/
Dominique

**Comment by mchobby** [comment]
This issue is now solved in version 1.6.0.20! 
Here some details about the solution.

The connector GPIOs have been reallocated in a different way!
![Sélection_143](https://user-images.githubusercontent.com/2701466/235550376-253637d0-78e0-4569-84e1-1816828c3be8.png)
With consecutive GPIO 26,27,28 we can bring a PIO_SPI bus. CS/ChipSelect signal is managed with GPIO 5.

Because of SPI bus, the I2C bus (as well as extra buzzer+USB_Power) **have been relocated** on GPIO 18 & 19.
As GPIOs 18 & 19 also support harware I2C (as I2C1) then the code change is minimal.

Finally, FatFs lib has been included and modified to support PIO_SPI instead of hardware SPI.

See:
* common/pio_spi : SPI bus over PIO. Version of the RPI foundation with a small change for SDCard support)
* common/pio_sd : only test the SDCard and send information on the debug line
* pio_fatfs/* : fatfs using the PIO_SPI implementation
* cli/* : small command line interpreter used to test SDCard features and lib.

Dom
